File: app/src/main/java/com/chichiangho/base/http/RxHttpDataProvider.java
package com.chichiangho.base.http;


import com.chichiangho.base.base.ProjectDefines;
import com.chichiangho.base.bean.User;
import com.chichiangho.base.utils.PreferenceUtil;
import com.chichiangho.base.utils.SignUtil;

import java.util.HashMap;

import io.reactivex.Observable;
import io.reactivex.ObservableEmitter;
import io.reactivex.ObservableOnSubscribe;

/**
 * Created by chichiangho on 2017/3/17.
 */

public class RxHttpDataProvider {

    private static String getUrl(String method) {
        return ProjectDefines.HTTP_BASE + ProjectDefines.URL_BASE + method;
    }

    /**
     * 构建请求参数
     *
     * @param bizParams
     * @param subSignParams
     * @return
     */
    private static RequestParam getRequestParam(BizParam bizParams, HashMap<String, String> subSignParams) {
        RequestParam requestParam = new RequestParam();
        requestParam.setAppKey(ProjectDefines.APP_KEY);
        requestParam.setTimestamp(System.currentTimeMillis());


        HashMap<String, String> signParams = new HashMap<>();
        signParams.put("appKey", requestParam.getAppKey());
        signParams.put("timestamp", requestParam.getTimestamp() + "");

        User user = PreferenceUtil.getUser();
        if (user != null) {
            requestParam.setToken(user.getUserToken());
            signParams.put("token", requestParam.getToken());

            requestParam.setUserId(user.getUserId());
            signParams.put("userId", requestParam.getUserId() + "");
        }

        if (!subSignParams.isEmpty())
            signParams.putAll(subSignParams);
        requestParam.setSign(SignUtil.getSign(signParams, ProjectDefines.APP_SECRET));

        requestParam.setParams(bizParams);

        return requestParam;
    }

    /*******************interfaces begin*********************/

    /**
     * 接口示例
     *
     * @param cellPhone
     */
    public static <T> Observable<T> example(final String cellPhone, final Class<T> clz) {

        return Observable.create(new ObservableOnSubscribe<T>() {
            @Override
            public void subscribe(ObservableEmitter<T> e) throws Exception {

                BizParam bizParams = new BizParam();
//                bizParams.setCellPhone(cellPhone);

                HashMap<String, String> signParams = new HashMap<>();
//                signParams.put("cellPhone", cellPhone);

                RequestParam requestParam = getRequestParam(bizParams, signParams);

                e.onNext(OkHttpClient.post(getUrl("notice/get_notice_info"), requestParam, clz));
                e.onComplete();//must call in every Observable,or onComplete() in Observer will not be called
            }
        }).flatMap(new HandleResponse<T>());
    }

}

File: app/src/main/java/com/chichiangho/base/http/RxHttpDataProvider.kt
package com.chichiangho.base.http


import com.chichaingho.library_rx_okhttp.HandleResponse
import com.chichaingho.library_rx_okhttp.OkHttpClient
import com.chichiangho.base.base.AppConfigs
import com.chichiangho.base.utils.Preference
import com.chichiangho.base.utils.SignUtil
import com.chichiangho.common.extentions.onNextComplete
import com.chichiangho.common.extentions.toJson
import io.reactivex.Observable
import io.reactivex.ObservableOnSubscribe
import java.util.*

/**
 * Created by chichiangho on 2017/3/17.
 */

object RxHttpDataProvider {

    private fun getUrl(method: String): String = AppConfigs.HTTP_BASE + AppConfigs.URL_BASE + method

    /**
     * 构建请求参数

     * @param bizParams
     * *
     * @param subSignParams
     * *
     * @return
     */
    private fun getRequestParam(bizParams: BizParam, subSignParams: HashMap<String, String>): RequestParam {
        val requestParam = RequestParam()
        requestParam.appKey = AppConfigs.APP_KEY
        requestParam.timestamp = System.currentTimeMillis()


        val signParams = HashMap<String, String>()
        signParams.put("appKey", requestParam.appKey as String)
        signParams.put("timestamp", requestParam.timestamp?.toString() ?: "")

        if (Preference.loginUser != null) {
            requestParam.token = Preference.loginUser?.userToken
            signParams.put("token", requestParam.token as String)

            requestParam.userId = Preference.loginUser?.userId
            signParams.put("userId", requestParam.userId?.toString() ?: "")
        }

        if (!subSignParams.isEmpty())
            signParams.putAll(subSignParams)
        requestParam.sign = SignUtil.getSign(signParams, AppConfigs.APP_SECRET)

        requestParam.params = bizParams

        return requestParam
    }

    /*******************interfaces begin */

    /**
     * 接口示例

     * @param cellPhone
     */
    fun <T> example(cellPhone: String, clz: Class<T>): Observable<T> {

        return Observable.create(ObservableOnSubscribe<T> { e ->
            val bizParams = BizParam()
            //                bizParams.setCellPhone(cellPhone);

            val signParams = HashMap<String, String>()
            //                signParams.put("cellPhone", cellPhone);

            val requestParam = getRequestParam(bizParams, signParams)

            e.onNextComplete(OkHttpClient.post(getUrl("notice/get_notice_info"), requestParam.toJson(), clz))
        }).map(HandleResponse<T>())
    }

}
